File: perl/problem119.pl
#!/usr/bin/env perl

use strict;
use warnings;

sub sum_digits {
    my ($num) = @_;
    my $sum = 0;

    while ($num) {
        $sum += $num % 10;
        $num /= 10;
    }
    return $sum;
}

sub main {
    my $cnt = 0;
    my $num_digits = 2;

    while (1) {
        my $max_num = 10 ** $num_digits;
        my $min_num = $max_num / 10;

        for my $digit_sum (2 .. $num_digits*9) {
            my $num = $digit_sum * $digit_sum;
            my $power = 2;

            while ($num < $max_num) {
                if ($num >= $min_num && $num < $max_num && $num == sum_digits($num) ** $power) {
                    return $num if ++$cnt == 30;
                }
                $num *= $digit_sum;
                $power++;
            }
        }
        $num_digits++;
    }
}

print(main, "\n");

File: python/problem119.py
#!/usr/bin/env python

"""Problem 119: Digit power sum"""

from itertools import count

from utils import sum_digits


def main():

    cnt = 0

    for num_digits in count(2):
        max_num = 10**num_digits
        min_num = max_num//10

        for digit_sum in range(2, num_digits*9 + 1):
            num = digit_sum * digit_sum
            p = 2

            while num < max_num:
                if min_num <= num < max_num and num == sum_digits(num)**p:
                    cnt += 1
                    if cnt == 30:
                        return num
                num *= digit_sum
                p += 1


if __name__ == "__main__":
    print(main())
